filters/include/pcl/filters/experimental/voxel_grid.h

```diff
@@ -0,0 +1,482 @@
+
+      if (save_leaf_layout_)
+        leaf_layout_[grid_it->first] = num_voxels_++;
+
+      return centroid;
+    }
+
+    return boost::none;
+  }
+
+  /** \brief The filter name. */
+  std::string filter_name_;
+
+  /** \brief The size of a leaf. */
+  Eigen::Vector4f leaf_size_ = Eigen::Vector4f::Zero();
+
+  /** \brief Internal leaf sizes stored as 1/leaf_size_ for efficiency reasons. */
+  Eigen::Array4f inverse_leaf_size_ = Eigen::Array4f::Zero();
+
+  /** \brief Set to true if leaf layout information needs to be saved in \a
+   * leaf_layout_. */
+  bool save_leaf_layout_ = false;
+
+  /** \brief The leaf layout information for fast access to cells relative to current
+   * position **/
+  std::vector<int> leaf_layout_;
+
+  /** \brief The minimum and maximum bin coordinates, the number of divisions, and the
+   * division multiplier. */
+  Eigen::Vector4i min_b_ = Eigen::Vector4i::Zero(), max_b_ = Eigen::Vector4i::Zero(),
+                  div_b_ = Eigen::Vector4i::Zero(), divb_mul_ = Eigen::Vector4i::Zero();
+
+  /** \brief The iterable grid object for storing information of each fraction of space
+   * in the filtering space defined by the grid */
+  Grid grid_;
+
+  GridFilterBase<VoxelStructT>* grid_filter_;
+
+  std::size_t num_voxels_;
+
+  bool downsample_all_data_;
+
+  std::size_t min_points_per_voxel_;
+};
+
+template <typename GridStruct, typename PointT = GET_POINT_TYPE(GridStruct)>
+class VoxelFilter : public GridFilterBase<GridStruct> {
+  using GridFilterBase<GridStruct>::getGridStruct;
+
+public:
+  /** \brief Set the voxel grid leaf size.
+   * \param[in] leaf_size the voxel grid leaf size
+   */
+  inline void
+  setLeafSize(const Eigen::Vector4f& leaf_size)
+  {
+    Eigen::Vector4f& leaf_size_ = getGridStruct().leaf_size_;
+    Eigen::Array4f& inverse_leaf_size_ = getGridStruct().inverse_leaf_size_;
+
+    leaf_size_ = leaf_size;
+    // Avoid division errors
+    if (leaf_size_[3] == 0)
+      leaf_size_[3] = 1;
+    // Use multiplications instead of divisions
+    inverse_leaf_size_ = 1 / leaf_size_.array();
+  }
+
+  /** \brief Set the voxel grid leaf size.
+   * \param[in] lx the leaf size for X
+   * \param[in] ly the leaf size for Y
+   * \param[in] lz the leaf size for Z
+   */
+  inline void
+  setLeafSize(const float lx, const float ly, const float lz)
+  {
+    Eigen::Vector4f& leaf_size_ = getGridStruct().leaf_size_;
+    Eigen::Array4f& inverse_leaf_size_ = getGridStruct().inverse_leaf_size_;
+
+    leaf_size_[0] = lx;
+    leaf_size_[1] = ly;
+    leaf_size_[2] = lz;
+    // Avoid division errors
+    if (leaf_size_[3] == 0)
+      leaf_size_[3] = 1;
+    // Use multiplications instead of divisions
+    inverse_leaf_size_ = 1 / leaf_size_.array();
+  }
+
+  /** \brief Get the voxel grid leaf size. */
+  inline Eigen::Vector3f
+  getLeafSize() const
+  {
+    const Eigen::Vector4f& leaf_size_ =
+        const_cast<VoxelFilter*>(this)->getGridStruct().leaf_size_;
+    return leaf_size_.head<3>();
+  }
```

Comment:
I tried to make the code compiles, but the only way I found is to have `const_cast<VoxelFilter*>(this)`

Reasons
1.  `this` is needed for accessing base method of a template class
2. `getLeafSize` is a const function, so `this` is somehow needed to be const too. Therefore I cast `this` to a const point

@kunaltyagi @mvieth Do you have any suggestion to avoid using `const_cast`?

Reply:
Did you try adding a const overload for `getGridStruct`?

Reply:
I didn't try `const GridStruct& GridStruct() const` before, now it works, thanks!

I create a new overload following this [stackoverflow answer](https://stackoverflow.com/a/123995/5486298)